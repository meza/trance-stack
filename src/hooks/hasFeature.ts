import { getVisitorIdFromRequest } from '~/session.server';
import splitClient from '~/split.server';
import type { Features } from '~/features';

/**
 * @name hasFeature
 * @description
 */
export const hasFeature = async (request: Request, feature: Features): Promise<boolean> => {
  await splitClient.ready();
  const visitorId = await getVisitorIdFromRequest(request);
  const serverTreatment = splitClient.getTreatment(visitorId, feature);

  return serverTreatment === 'on';
};

