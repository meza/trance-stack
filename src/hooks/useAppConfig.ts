import { useOutletContext } from '@remix-run/react';

export const useAppConfig = <K extends keyof AppConfig>(key: K) =>
  useOutletContext<{ appConfig: AppConfig }>().appConfig[key];
