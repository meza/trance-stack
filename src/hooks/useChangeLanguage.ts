import { useEffect } from 'react';
import { useTranslation } from 'react-i18next';

/**
 * Change language polyfill
 * Required due to a bug in react-i18next
 * @see https://github.com/sergiodxa/remix-i18next/issues/107
 * @param locale{string} - locale to change to
 */
export const useChangeLanguage = (locale: string) => {
  const { i18n } = useTranslation();
  useEffect(() => {
    i18n.changeLanguage(locale);
  }, [locale, i18n]);
};
