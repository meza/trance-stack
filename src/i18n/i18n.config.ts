import type { InitOptions } from 'i18next';

export const supportedLngs = ['en'];
export const defaultNS = ['translation'];
export const fallbackLng = 'en';

const baseConfig: InitOptions = {
  supportedLngs: supportedLngs,
  fallbackLng: fallbackLng,
  defaultNS: defaultNS
};

export const getBaseClientConfig = (): InitOptions => ({
  supportedLngs: baseConfig.supportedLngs,
  fallbackLng: baseConfig.fallbackLng,
  defaultNS: baseConfig.defaultNS,
  interpolation: { escapeValue: false },
  react: { useSuspense: false },
  detection: {
    order: ['htmlTag'],
    caches: []
  }
});

export default baseConfig;
