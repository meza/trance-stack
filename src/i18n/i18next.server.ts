import { resolve } from 'node:path';
import { createInstance } from 'i18next';
import Backend from 'i18next-fs-backend';
import { initReactI18next } from 'react-i18next';
import baseConfig from './i18n.config';
import remixI18next from './remix.server';
import type { EntryContext } from '@remix-run/node';
import type { i18n } from 'i18next';

let i18nextInstance: i18n;

const init = async (
  locale: string,
  remixContext: EntryContext
): Promise<i18n> => {
  if (!i18nextInstance) {
    i18nextInstance = createInstance().use(initReactI18next).use(Backend);
    await i18nextInstance.init({
      ...baseConfig,
      lng: locale,
      ns: remixI18next.getRouteNamespaces(remixContext),
      backend: {
        loadPath: resolve('../public/locales/{{lng}}/{{ns}}.json')
      }
    });
  }
  return i18nextInstance;
};

export default init;
