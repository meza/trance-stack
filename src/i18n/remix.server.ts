import { resolve } from 'node:path';
import Backend from 'i18next-fs-backend';
import { RemixI18Next } from 'remix-i18next';
import { supportedLngs, fallbackLng } from './i18n.config';
import type { NewableModule } from 'i18next';

const i18next = new RemixI18Next({
  detection: {
    supportedLanguages: supportedLngs,
    fallbackLanguage: fallbackLng
  },
  i18next: {
    backend: {
      loadPath: resolve('../public/locales/{{lng}}/{{ns}}.json')
    }
  },
  backend: Backend as NewableModule<Backend>
});

export default i18next;
