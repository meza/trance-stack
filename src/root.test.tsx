import React, { useContext } from 'react';
import { useLoaderData } from '@remix-run/react';
import { render } from '@testing-library/react';
import { useTranslation } from 'react-i18next';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import App, { handle, links, loader, meta } from './root';
import { useChangeLanguage } from '~/hooks/useChangeLanguage';
import { remixI18next } from '~/i18n';
import { createUserSession } from '~/session.server';

vi.mock('~/session.server');
vi.mock('~/i18n');
vi.mock('./styles/app.css', () => ({ default: 'app.css' }));
vi.mock('./styles/dark.css', () => ({ default: 'dark.css' }));
vi.mock('./styles/light.css', () => ({ default: 'light.css' }));
vi.mock('@remix-run/react');
vi.mock('react-i18next');
vi.mock('~/hooks/useChangeLanguage');
vi.mock('~/components/Posthog', () => ({
  Posthog: () => <div id={'mock-posthog'}/>
}));
vi.mock('~/components/CookieConsent', async () => {
  const actual = await vi.importActual('~/components/CookieConsent') as object;
  return {
    ...actual,
    CookieConsentBanner: () => <div id={'mock-cookie-consent-banner'}/>
  };
});
vi.mock('react', async () => ({
  // eslint-disable-next-line @typescript-eslint/consistent-type-imports
  ...await vi.importActual<typeof import('react')>('react'),
  useContext: vi.fn()
}));

vi.mock('@sentry/remix', () => ({
  withSentry: (Component: React.FC) => {
    return () => {
      return (
        <>
          {'mock sentry wrapper'}
          <Component/>
        </>
      );
    };
  }
}));

describe('The root module', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('should have a consistent meta function', () => {
    expect(meta({} as never)).toMatchInlineSnapshot(`
      {
        "charset": "utf-8",
        "title": "REPL_APP_NAME",
        "viewport": "width=device-width,initial-scale=1",
      }
    `);
  });

  it('should have a consistent links function', () => {
    expect(links()).toMatchInlineSnapshot(`
      [
        {
          "href": "app.css",
          "rel": "stylesheet",
        },
        {
          "href": "/_static/favicon.ico",
          "rel": "icon",
          "type": "image/x-icon",
        },
      ]
    `);
  });

  it('should expose a consistent handle', () => {
    expect(handle).toMatchInlineSnapshot(`
      {
        "i18n": [
          "translation",
        ],
      }
    `);
  });

  describe('when calling the loader', () => {
    beforeEach(() => {
      vi.mocked(createUserSession).mockResolvedValue({
        visitorId: 'a-visitorId',
        cookie: 'a-cookie',
        session: {
          get: vi.fn(),
          set: vi.fn()
        } as never
      });
      vi.mocked(remixI18next.getLocale).mockResolvedValue('en');

      vi.stubEnv('NODE_ENV', 'development');
      vi.stubEnv('GOOGLE_ANALYTICS_ID', 'ga-id');
      vi.stubEnv('HOTJAR_ID', 'a-hotjar-id');
      vi.stubEnv('MIXPANEL_TOKEN', 'a-mixpanel-token');
      vi.stubEnv('MIXPANEL_API', 'a-mixpanel-api');
      vi.stubEnv('COOKIEYES_TOKEN', 'a-cookieyes-token');

    });

    it('should return the app config for dev', async () => {
      const request = new Request('https://example.com');
      const response = await loader({ request: request } as never);
      const data = await response.json();
      expect(data).toMatchInlineSnapshot(`
        {
          "appConfig": {
            "cookieYesToken": "a-cookieyes-token",
            "googleAnalyticsId": "ga-id",
            "hotjarId": "a-hotjar-id",
            "isProduction": false,
            "version": "0.0.0-dev",
            "visitorId": "a-visitorId",
          },
          "locale": "en",
        }
      `);
    });

    it('should set the cookie header', async () => {
      const request = new Request('https://example.com');
      const response = await loader({ request: request } as never);
      expect(response.headers.get('Set-Cookie')).toMatchInlineSnapshot('"a-cookie"');
    });

    it('should return the app config for prod', async () => {
      vi.stubEnv('NODE_ENV', 'production');
      const request = new Request('https://example.com');
      const response = await loader({ request: request } as never);
      const data = await response.json();
      expect(data).toMatchInlineSnapshot(`
        {
          "appConfig": {
            "cookieYesToken": "a-cookieyes-token",
            "googleAnalyticsId": "ga-id",
            "hotjarId": "a-hotjar-id",
            "isProduction": true,
            "version": "0.0.0-dev",
            "visitorId": "a-visitorId",
          },
          "locale": "en",
        }
      `);
    });

    it('should prepare the environment', async () => {
      const request = new Request('https://example.com');
      await loader({ request: request } as never);

      expect(createUserSession).toHaveBeenCalledWith(request);
      expect(remixI18next.getLocale).toHaveBeenCalledWith(request);

    });
  });

  describe('when rendering the app', () => {
    const appConfig: AppConfig = {
      hotjarId: 'a-hotjar-id',
      googleAnalyticsId: 'ga-id',
      visitorId: 'a-visitor-id',
      isProduction: true,
      cookieYesToken: 'a-cookieyes-token',
      version: '0.0.0-dev',
      sentryDsn: 'a-sentry-dsn',
      posthogToken: 'a-posthog-token',
      posthogApi: 'a-posthog-api'
    };

    beforeEach(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore - weird stuff happening with the useTranslataion mocks
      vi.mocked(useTranslation).mockReturnValue({
        i18n: {
          language: 'en',
          dir: () => 'ltr'
        }
      } as never);
      vi.mocked(useChangeLanguage).mockReturnValue();
    });

    describe('when there is no consent for cookies', () => {
      beforeEach(() => {
        vi.mocked(useLoaderData).mockReturnValue({ appConfig: appConfig, locale: 'en' } as never);
      });
      it('renders the app', () => {
        vi.mocked(useContext).mockReturnValue('mocked-nonce');
        const errorSpy = vi.spyOn(console, 'error').mockImplementation(() => {
          // There is a DOM Nesting Validation error because we're rendering
          // the entire html in a test environment. We don't care about that error
        });
        const markup = render(<App/>);
        errorSpy.mockReset();
        expect(markup.asFragment()).toMatchSnapshot();
        expect(markup.getByText('mock sentry wrapper')).toBeInTheDocument();
      });
    });

    describe('when there is consent for cookies', () => {
      beforeEach(() => {
        vi.mocked(useLoaderData).mockReturnValue({
          appConfig: appConfig,
          locale: 'en',
          colorMode: 'dark',
          consentData: {
            analytics: true
          }
        } as never);
      });
      it('renders the app with analytics scripts', () => {
        vi.mocked(useContext).mockReturnValue('mocked-nonce');
        const errorSpy = vi.spyOn(console, 'error').mockImplementation(() => {
          // There is a DOM Nesting Validation error because we're rendering
          // the entire html in a test environment. We don't care about that error
        });
        const markup = render(<App/>);
        errorSpy.mockReset();
        expect(markup.asFragment()).toMatchSnapshot();
      });
    });
  });
});
