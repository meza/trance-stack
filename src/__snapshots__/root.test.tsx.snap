// Vitest Snapshot v1

exports[`The root module > when rendering the app > when there is consent for cookies > renders the app with analytics scripts 1`] = `
<DocumentFragment>
  mock sentry wrapper
  <script
    id="app-config"
    nonce="mocked-nonce"
  >
    window.appConfig = {"hotjarId":"a-hotjar-id","googleAnalyticsId":"ga-id","mixpanelToken":"a-mixpanel-token","visitorId":"a-visitor-id","isProduction":true,"mixpanelApi":"a-mixpanel-api","splitToken":"a-split-token","cookieYesToken":"a-cookieyes-token","version":"0.0.0-dev","sentryDsn":"a-sentry-dsn"}
  </script>
  <script
    id="color-mode-sensor"
    nonce="mocked-nonce"
  >
    (() =&gt; {
  const cl = document.firstElementChild?.classList;
  if (!(cl?.contains("dark") || cl?.contains("light"))) {
    cl?.add(window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light");
  }
})();
  </script>
  <script
    id="color-mode-update"
    nonce="mocked-nonce"
  >
    (() =&gt; {
  addEventListener("load", () =&gt; {
    window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", (e) =&gt; {
      const cl = document.firstElementChild?.classList;
      cl?.remove(e.matches ? "light" : "dark");
      cl?.add(e.matches ? "dark" : "light");
    });
  });
})();
  </script>
  <script
    id="google-analytics"
    nonce="mocked-nonce"
  >
    window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'ga-id', { user_id: 'a-visitor-id' });
  </script>
  <script
    async=""
    id="hotjar-init"
    nonce="mocked-nonce"
  >
    (function(h){
            h.hj=h.hj||function(){(h.hj.q=h.hj.q||[]).push(arguments)};
            h._hjSettings={hjid:'a-hotjar-id',hjsv:6,hjdebug:false};
          })(window);
          hj('identify', 'a-visitor-id');
          
  </script>
</DocumentFragment>
`;

exports[`The root module > when rendering the app > when there is no consent for cookies > renders the app 1`] = `
<DocumentFragment>
  mock sentry wrapper
  <script
    id="app-config"
    nonce="mocked-nonce"
  >
    window.appConfig = {"hotjarId":"a-hotjar-id","googleAnalyticsId":"ga-id","mixpanelToken":"a-mixpanel-token","visitorId":"a-visitor-id","isProduction":true,"mixpanelApi":"a-mixpanel-api","splitToken":"a-split-token","cookieYesToken":"a-cookieyes-token","version":"0.0.0-dev","sentryDsn":"a-sentry-dsn"}
  </script>
  <script
    id="color-mode-sensor"
    nonce="mocked-nonce"
  >
    (() =&gt; {
  const cl = document.firstElementChild?.classList;
  if (!(cl?.contains("dark") || cl?.contains("light"))) {
    cl?.add(window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light");
  }
})();
  </script>
  <script
    id="color-mode-update"
    nonce="mocked-nonce"
  >
    (() =&gt; {
  addEventListener("load", () =&gt; {
    window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", (e) =&gt; {
      const cl = document.firstElementChild?.classList;
      cl?.remove(e.matches ? "light" : "dark");
      cl?.add(e.matches ? "dark" : "light");
    });
  });
})();
  </script>
  <script
    id="google-analytics"
    nonce="mocked-nonce"
  >
    window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'ga-id', { user_id: 'a-visitor-id' });
  </script>
  <script
    async=""
    id="hotjar-init"
    nonce="mocked-nonce"
  >
    (function(h){
            h.hj=h.hj||function(){(h.hj.q=h.hj.q||[]).push(arguments)};
            h._hjSettings={hjid:'a-hotjar-id',hjsv:6,hjdebug:false};
          })(window);
          hj('identify', 'a-visitor-id');
          
  </script>
  <div
    id="mock-cookie-consent-banner"
  />
</DocumentFragment>
`;
